### ISSUE TITLE
Model Logic | Why doesn't the score go down?

### ISSUE THREAD

**Issue by oxillix** [AICoE/log-anomaly-detector#192]
Hi,

First and foremost, Thank you guys for this amazing project!

1. I started training the model on the Hadoop_2k test dataset with this command:
pipenv run python3 app.py run --job-type train

2. After the model was trained I changed the same dataset, and I added some obvious anomaly lines like: "This is an anomaly fo sure!!!!", note that i kept the rest of the file like it was.

3. I ran an inference job on the changed dataset with this command:
pipenv run python3 app.py run --job-type train

4. **I checked the local-results.txt file and i noticed something I don't really understand.**
```
[{"message": "INFO [main] org.apache.hadoop.mapreduce.v2.app.MRAppMaster: This logline is sooooooooooooo anomaly this does not belong here at all", "predict_id": "79761d6e-13b0-4a2b-a428-86b5c8025d3e", "anomaly_score": 0.9355214564062759, "anomaly": 0},
{"message": "INFO [main] org.apache.hadoop.mapreduce.v2.app.MRAppMaster: Created MRAppMaster for application appattempt_1445144423722_0020_000001", "predict_id": "9772a16d-057a-48f7-90f6-6f2575f56758", "anomaly_score": 0.9266197762630918, "anomaly": 0},
{"message": "INFO [main] org.apache.hadoop.mapreduce.v2.app.MRAppMaster: Executing with tokens:", "predict_id": "de93e113-5139-4e68-b20b-09d5a83d9b19", "anomaly_score": 0.9413419113855845, "anomaly": 0}
```
**I expected that the previously trained loglines would have a very low anomaly_score and my newly added logline would have a very high score.** This wasnt the case. Am i missing something? Does the algorithm work in a different way?

**Comment by zmhassan** [comment]
@MichaelClifford could you share some details about the scoring.

**Comment by MichaelClifford** [comment]
Hi,  Thanks for using the project! 

At this stage this kinda of feedback is super helpful for us to improve on the project and ensure we are building a useful tool. : ) 

Your intuition is correct, that we should expect previously unseen or lower frequency log lines to have a  higher anomaly score, as the scoring mechanism is currently to take the L2 norm of the test log from its best matching unit on the trained Self-Organizing Map. With this structure we would hope to see anomalous logs be further away from nodes on the map and therefore have a higher anomaly score. That said, perhaps switching to a cosine distance metric would yield better results here.  

There are also a few factors worth considering that could muddle this higher score expectation. The word2vec encoding and its associated hyper-parameters can have an effect on the logs encoded values. Ideally, it finds an encoding scheme that separates “normal” logs from “anomalous” ones in the encoded space, but since it relies on some deep learning mechanisms to generate the encoding, interpretability can become an issue and the model may encode logs to similar vectors for reasons that might not be immediately obvious to a human user. So making sure that we have the right window size, vector length and number of training epochs for our training set can have a significant impact on the resulting encoded log-space.        

Also, this particular dataset is quite small, 2000 logs in total and after cleaning there are only 141 unique entries. With a default map size of 24x24 (or 576 nodes), its not clear to me that there couldn’t be nodes left on the map that aren’t representative of the training data at all, confounding the distance metrics. So, for this dataset a SOM with map_size  of 10 or smaller might be more appropriate.  

This is all to say, maybe a different dataset along with some additional parameter tuning, specifically around the word2vec encoding and the size of the SOM map should generate more consistent and expected results.

Thanks for bring this issue to our attention, I think I will add a demo or notebook of some kind to the repo for users to walk through and get a better sense for how the tool should work and what potential pitfalls might present themselves. 

Thanks so much, and I hope that was helpful.

**Comment by zmhassan** [comment]
We are assuming that this question is answered. Feel free to reopen this if you have further questions.
